core/src/main/kotlin/com/strumenta/kolasu/transformation/Transformation.kt

```diff
@@ -17,15 +20,26 @@ annotation class Mapped(val path: String = "")
 /**
  * Factory that, given a tree node, will instantiate the corresponding transformed node.
  */
-class NodeFactory<Source>(
-    val constructor: (Source, ASTTransformer, NodeFactory<Source>) -> Node,
+class NodeFactory<Source, Output : Node>(
+    val constructor: (Source, ASTTransformer, NodeFactory<Source, Output>) -> Output,
     val children: MutableMap<String, ChildNodeFactory<Source, *, *>?> = mutableMapOf()
 ) {
+
+    private var finalizer: (Output) -> Unit = {}
+
+    fun finally(node: Output) {
+        this.finalizer(node)
+    }
+
+    fun finally(finalizer: (Output) -> Unit) {
```

Comment:
Should this be a property?

Reply:
The reason why I splitted `finally` and `finalizer` is to allow users to define the `finalizer` in a more fluid way. That is:
```kotlin
registerNodeFactory(...)
    .withChild(...)
    .finally { ... }
```
rather than
```kotlin
registerNodeFactory(...)
    .withChild(...)
    .apply { this.finalizer = {...} }
```
or 
```
val nodeFactory = registerNodeFactory(...).withChild(...)
nodeFactory.finalizer = {...}
```
Does it make sense? Maybe a more explicit name like `onAllChildrenProcessed` could work better than `finalizer` and `finally` though.